File: 129-sum-root-to-leaf-numbers.ts
function sumNumbers(root: TreeNode | null) {
  return sumNumbersIterative(root)
}

function sumNumbersRecursive(
  root: TreeNode | null,
  sequence: number = 0
): number {
  if (!root) {
    return 0
  }
  const isLeaf = !root.left && !root.right
  const num = sequence * 10 + root.val
  if (isLeaf) {
    return num
  } else {
    const leftNum = root.left ? sumNumbersRecursive(root.left, num) : 0
    const rightNum = root.right ? sumNumbersRecursive(root.right, num) : 0
    return leftNum + rightNum
  }
}

function sumNumbersIterative(root: TreeNode | null): number {
  if (!root) {
    return 0
  }
  let sum = 0
  const stack: [TreeNode, number][] = [[root, 0]]

  while (stack.length) {
    const [node, acc] = stack.pop()!
    const isLeaf = !node.left && !node.right
    const val = acc * 10 + node.val
    if (isLeaf) {
      sum += val
    } else {
      if (node.left) {
        stack.push([node.left, val])
      }
      if (node.right) {
        stack.push([node.right, val])
      }
    }
  }

  return sum
}

// There is another solution called the Morris Preorder Traversal but it's nuanced and
// I'm not going to bother memorizing it. It allows for use of no extra space.

File: 129_sum_root_to_leaf_numbers.py
from __future__ import annotations
from dataclasses import dataclass


@dataclass
class TreeNode:
    val: int
    left: TreeNode | None
    right: TreeNode | None


class Solution:
    def sumNumbers(self, root: TreeNode | None) -> int:
        return self.sumNumbersIterative(root)

    def sumNumbersRecursive(self, root: TreeNode | None, acc: int = 0) -> int:
        if not root:
            return 0
        is_leaf = root.left is None and root.right is None
        val = acc * 10 + root.val
        if is_leaf:
            return val
        else:
            left_val = self.sumNumbersRecursive(root.left, val) if root.left else 0
            right_val = self.sumNumbersRecursive(root.right, val) if root.right else 0
            return left_val + right_val

    def sumNumbersIterative(self, root: TreeNode | None) -> int:
        if not root:
            return 0
        sum = 0
        stack = [(root, 0)]
        while stack:
            node, acc = stack.pop()
            is_leaf = node.left is None and node.right is None
            val = acc * 10 + node.val
            if is_leaf:
                sum += val
            else:
                if node.left:
                    stack.append((node.left, val))
                if node.right:
                    stack.append((node.right, val))
        return sum
